File: util/src/main/java/dev/openrs2/util/collect/DisjointSet.java
package dev.openrs2.util.collect;

public interface DisjointSet<T> extends Iterable<DisjointSet.Partition<T>> {
	interface Partition<T> extends Iterable<T> {
		/* empty */
	}

	Partition<T> add(T x);
	Partition<T> get(T x);
	void union(Partition<T> x, Partition<T> y);
	int elements();
	int partitions();
}

File: util/src/main/java/dev/openrs2/util/collect/DisjointSet.kt
package dev.openrs2.util.collect

/**
 * A data structure containing a set of elements partitioned into a number of
 * non-overlapping subsets. New elements belong to singleton subsets. The
 * [union] function combines two subsets together into a single larger subset.
 */
public interface DisjointSet<T> : Iterable<DisjointSet.Partition<T>> {
    public interface Partition<T> : Iterable<T>

    public val elements: Int
    public val partitions: Int

    public fun add(x: T): Partition<T>
    public operator fun get(x: T): Partition<T>?
    public fun union(x: Partition<T>, y: Partition<T>)
}
